cwl_utils/cwl_v1_2_expression_refactor.py

```diff
@@ -1900,36 +1904,17 @@ def traverse_step(
                 target = get_input_for_id(inp.id, original_process)
                 if not target:
                     raise WorkflowException("target not found")
-                input_source_id = None
+                input_source_id: Optional[Union[str, List[str]]] = None
                 source_type: Optional[
                     Union[List[cwl.WorkflowInputParameter], cwl.WorkflowInputParameter]
                 ] = None
                 if inp.source:
-                    if isinstance(inp.source, MutableSequence):
-                        input_source_id = []
-                        source_types: List[cwl.WorkflowInputParameter] = []
-                        for source in inp.source:
-                            source_id = source.split("#")[-1]
-                            input_source_id.append(source_id)
-                            temp_type = utils.type_for_source(
-                                step.run, source_id, parent
-                            )
-                            if isinstance(temp_type, list):
-                                for ttype in temp_type:
-                                    if ttype not in source_types:
-                                        source_types.append(ttype)
-                            else:
-                                if temp_type not in source_types:
-                                    source_types.append(temp_type)
-                        source_type = cwl.WorkflowInputParameter(
-                            id=None,
-                            type=cwl.ArraySchema(source_types, "array"),
-                        )
-                    else:
-                        input_source_id = inp.source.split("#")[-1]
-                        source_type = utils.param_for_source_id(
-                            step.run, input_source_id, parent
-                        )
+                    input_source_id = (
+                        [source.split("#")[-1] for source in inp.source]
+                        if isinstance(inp.source, MutableSequence)
+                        else inp.source.split("#")[-1]
+                    )
+                    source_type = utils.type_for_source(step.run, inp.source, parent)
```

Comment:
How is this working? `source_type` must be a `cwl.WorkflowInputParameter` but that is not possible from the proposed `utils.type_for_source`.